File: backend/src/main/java/com/auto/trader/trade/IndicatorUtil.java
package com.auto.trader.trade;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

import com.auto.trader.trade.dto.CandleDto;

import lombok.AllArgsConstructor;
import lombok.Getter;

public class IndicatorUtil {
	  
	   @Getter
	   public static class IndicatorPoint {
	        private long time;
	        private Double value;
	        public IndicatorPoint(long time, Double value) {
	            this.time = time;
	            this.value = value;
	        }
	    }
	    @Getter
	    public static class DualIndicatorPoint {
	        private long time;
	        private Double k;
	        private Double d;
	        // getters/setters
	        public DualIndicatorPoint(long time, Double k, Double d) {
	            this.time = time;
	            this.k = k;
	            this.d = d;
	        }
	    }
	    @Getter
	    public static class VWBB {
	        private List<IndicatorPoint> upper;
	        private List<IndicatorPoint> lower;
	        private List<IndicatorPoint> basis;
	        // getters/setters
	        private VWBB(List<IndicatorPoint> upper, List<IndicatorPoint> lower, List<IndicatorPoint> basis) {
	        	this.upper = upper;
	        	this.lower = lower;
	        	this.basis = basis;
	        }
	    }

	public static List<IndicatorPoint> calculateRSI(List<CandleDto> candles, int period) {
		List<IndicatorPoint> rsiPoints = new ArrayList<>();
		if (candles.size() < period + 1)
			return rsiPoints;

		double gainSum = 0;
		double lossSum = 0;

		for (int i = 1; i <= period; i++) {
			double diff = candles.get(i).getClose() - candles.get(i - 1).getClose();
			if (diff >= 0)
				gainSum += diff;
			else
				lossSum += -diff;
		}

		double avgGain = gainSum / period;
		double avgLoss = lossSum / period;
		double rs = avgLoss == 0 ? 100 : avgGain / avgLoss;
		rsiPoints.add(new IndicatorPoint(candles.get(period).getTime(), 100 - (100 / (1 + rs))));

		for (int i = period + 1; i < candles.size(); i++) {
			double diff = candles.get(i).getClose() - candles.get(i - 1).getClose();
			double gain = diff > 0 ? diff : 0;
			double loss = diff < 0 ? -diff : 0;

			avgGain = ((avgGain * (period - 1)) + gain) / period;
			avgLoss = ((avgLoss * (period - 1)) + loss) / period;
			rs = avgLoss == 0 ? 100 : avgGain / avgLoss;

			double rsi = 100 - (100 / (1 + rs));
			rsiPoints.add(new IndicatorPoint(candles.get(i).getTime(), rsi));
		}

		return rsiPoints;
	}

	public static List<DualIndicatorPoint> calculateStochRSI(List<CandleDto> candles, int period, int signalPeriod) {
		List<IndicatorPoint> rsi = calculateRSI(candles, period);
		List<Double> stochK = new ArrayList<>(Collections.nCopies(rsi.size(), null));

		for (int i = period * 2; i < rsi.size(); i++) {
			List<Double> window = new ArrayList<>();
			for (int j = i - period + 1; j <= i; j++) {
				Double v = rsi.get(j).getValue();
				if (v != null)
					window.add(v);
			}

			if (window.size() < period)
				continue;

			double min = Collections.min(window);
			double max = Collections.max(window);
			Double current = rsi.get(i).getValue();

			if (current != null && max != min) {
				stochK.set(i, (current - min) / (max - min) * 100);
			}
		}

		List<DualIndicatorPoint> result = new ArrayList<>();
		for (int i = 0; i < rsi.size(); i++) {
			Double k = stochK.get(i);
			Double dVal = null;

			if (k != null && i >= signalPeriod) {
				List<Double> recentK = stochK.subList(i - signalPeriod + 1, i + 1).stream().filter(Objects::nonNull)
						.collect(Collectors.toList());

				if (recentK.size() == signalPeriod) {
					dVal = recentK.stream().mapToDouble(v -> v).average().orElse(0);
				}
			}

			result.add(new DualIndicatorPoint(rsi.get(i).getTime(), k, dVal));
		}

		return result;
	}

	public static List<IndicatorPoint> calculateEMA(List<CandleDto> candles, int period) {
		List<IndicatorPoint> result = new ArrayList<>();
		if (candles.isEmpty())
			return result;

		double k = 2.0 / (period + 1);
		double emaPrev = candles.get(0).getClose();

		for (int i = 0; i < candles.size(); i++) {
			CandleDto c = candles.get(i);
			if (i == 0) {
				result.add(new IndicatorPoint(c.getTime(), emaPrev));
			} else {
				emaPrev = c.getClose() * k + emaPrev * (1 - k);
				result.add(new IndicatorPoint(c.getTime(), emaPrev));
			}
		}

		return result;
	}

	public static List<IndicatorPoint> calculateSMA(List<CandleDto> candles, int period) {
		List<IndicatorPoint> result = new ArrayList<>();

		for (int i = 0; i < candles.size(); i++) {
			if (i < period - 1) {
				result.add(new IndicatorPoint(candles.get(i).getTime(), null));
			} else {
				double sum = 0;
				for (int j = i - period + 1; j <= i; j++) {
					sum += candles.get(j).getClose();
				}
				double sma = sum / period;
				result.add(new IndicatorPoint(candles.get(i).getTime(), sma));
			}
		}

		return result;
	}
	

	public static VWBB calculateVWBB(List<CandleDto> candles, int period, double multiplier) {
		int size = candles.size();
		List<Double> vwma = new ArrayList<>(Collections.nCopies(size, null));
		List<Double> std = new ArrayList<>(Collections.nCopies(size, null));

		for (int i = period - 1; i < size; i++) {
			double volSum = 0, priceVolSum = 0;
			for (int j = i - period + 1; j <= i; j++) {
				double price = candles.get(j).getClose();
				double volume = candles.get(j).getVolume() != 0 ? candles.get(j).getVolume() : 1;
				volSum += volume;
				priceVolSum += price * volume;
			}
			vwma.set(i, priceVolSum / volSum);
		}

		for (int i = period - 1; i < size; i++) {
			if (vwma.get(i) == null)
				continue;
			double mean = vwma.get(i);
			double sum = 0;
			for (int j = i - period + 1; j <= i; j++) {
				double diff = candles.get(j).getClose() - mean;
				sum += diff * diff;
			}
			std.set(i, Math.sqrt(sum / period));
		}

		List<IndicatorPoint> basis = new ArrayList<>();
		List<IndicatorPoint> upper = new ArrayList<>();
		List<IndicatorPoint> lower = new ArrayList<>();

		for (int i = 0; i < size; i++) {
			long time = candles.get(i).getTime();
			Double mean = vwma.get(i);
			Double deviation = std.get(i);

			basis.add(new IndicatorPoint(time, mean));
			upper.add(new IndicatorPoint(time,
					(mean != null && deviation != null) ? mean + multiplier * deviation : null));
			lower.add(new IndicatorPoint(time,
					(mean != null && deviation != null) ? mean - multiplier * deviation : null));
		}

		return new VWBB(upper, lower, basis);
	}

}

File: frontend/src/shared/util/indicatorUtil.ts
// 파일: src/shared/util/indicatorUtil.ts

export interface IndicatorPoint {
  time: number;
  value: number | null;
}

export interface DualIndicatorPoint {
  time: number;
  k: number | null;
  d: number | null;
}

export interface Candle {
  time: number;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export const calculateEMA = (data: Candle[], period = 20): IndicatorPoint[] => {
  const k = 2 / (period + 1);
  let emaPrev = data[0].close;

  return data.map((d, i) => {
    if (i === 0) return { time: d.time, value: emaPrev };
    emaPrev = d.close * k + emaPrev * (1 - k);
    return { time: d.time, value: emaPrev };
  });
};

export function calculateSMA(
  data: IndicatorPoint[],
  period: number
): IndicatorPoint[] {
  const result: IndicatorPoint[] = [];

  for (let i = 0; i < data.length; i++) {
    if (i < period - 1 || data[i].value == null) {
      result.push({ time: data[i].time, value: null });
      continue;
    }

    const window = data.slice(i - period + 1, i + 1);
    const valid = window.filter((d) => typeof d.value === "number");

    if (valid.length < period) {
      result.push({ time: data[i].time, value: null });
      continue;
    }

    const avg = valid.reduce((sum, d) => sum + d.value!, 0) / valid.length;

    result.push({ time: data[i].time, value: avg });
  }

  return result;
}

export const calculateRSI = (data: Candle[], period = 14): IndicatorPoint[] => {
  let avgGain = 0;
  let avgLoss = 0;
  const result: IndicatorPoint[] = [];

  for (let i = 1; i < data.length; i++) {
    const diff = data[i].close - data[i - 1].close;
    const gain = diff > 0 ? diff : 0;
    const loss = diff < 0 ? -diff : 0;

    if (i <= period) {
      avgGain += gain;
      avgLoss += loss;
      result.push({ time: data[i].time, value: null });
    } else if (i === period + 1) {
      avgGain = avgGain / period;
      avgLoss = avgLoss / period;
      const rs = avgLoss === 0 ? 100 : avgGain / avgLoss;
      const rsi = 100 - 100 / (1 + rs);
      result.push({ time: data[i].time, value: rsi });
    } else {
      avgGain = (avgGain * (period - 1) + gain) / period;
      avgLoss = (avgLoss * (period - 1) + loss) / period;
      const rs = avgLoss === 0 ? 100 : avgGain / avgLoss;
      const rsi = 100 - 100 / (1 + rs);
      result.push({ time: data[i].time, value: rsi });
    }
  }

  return result;
};

export function calculateStochRSI(
  data: Candle[],
  rsiPeriod = 14,
  stochPeriod = 14,
  kPeriod = 3,
  dPeriod = 3
): DualIndicatorPoint[] {
  const rsi = calculateRSI(data, rsiPeriod);
  const stochRsi: IndicatorPoint[] = [];

  for (let i = 0; i < rsi.length; i++) {
    if (i < stochPeriod - 1) {
      stochRsi.push({ time: rsi[i].time, value: null });
      continue;
    }

    const sliced = rsi.slice(i - stochPeriod + 1, i + 1);
    const minRsi = Math.min(...sliced.map((d) => d.value!));
    const maxRsi = Math.max(...sliced.map((d) => d.value!));

    const denominator = maxRsi - minRsi;
    if (denominator === 0) {
      stochRsi.push({ time: rsi[i].time, value: 0 });
    } else {
      const value = ((rsi[i].value! - minRsi) / denominator) * 100;
      stochRsi.push({ time: rsi[i].time, value });
    }
  }

  // ✅ Binance 기준으로 smoothing: SMA 사용
  const smoothedK = calculateSMA(stochRsi, kPeriod);
  const smoothedD = calculateSMA(smoothedK, dPeriod);

  const result: DualIndicatorPoint[] = [];
  for (let i = 0; i < stochRsi.length; i++) {
    result.push({
      time: stochRsi[i].time,
      k: smoothedK[i]?.value ?? null,
      d: smoothedD[i]?.value ?? null,
    });
  }

  return result;
}

export const calculateVWBB = (
  candles: Candle[],
  period = 20,
  multiplier = 2
): {
  upper: IndicatorPoint[];
  lower: IndicatorPoint[];
  basis: IndicatorPoint[];
} => {
  const vwma = candles.map((_, i) => {
    if (i < period - 1) return null;
    let volSum = 0;
    let priceVolSum = 0;
    for (let j = i - period + 1; j <= i; j++) {
      const price = candles[j].close;
      const volume = candles[j].volume ?? 1; // ✅ volumeWeightRatio 제거
      volSum += volume;
      priceVolSum += price * volume;
    }
    return priceVolSum / volSum;
  });

  const std = candles.map((_, i) => {
    if (i < period - 1 || vwma[i] == null) return null;
    const mean = vwma[i]!;
    const variance =
      candles
        .slice(i - period + 1, i + 1)
        .reduce((sum, d) => sum + Math.pow(d.close - mean, 2), 0) / period;
    return Math.sqrt(variance);
  });

  return {
    basis: vwma.map((v, i) => ({ time: candles[i].time, value: v })),
    upper: vwma.map((v, i) =>
      v == null || std[i] == null
        ? { time: candles[i].time, value: null }
        : { time: candles[i].time, value: v + multiplier * std[i]! }
    ),
    lower: vwma.map((v, i) =>
      v == null || std[i] == null
        ? { time: candles[i].time, value: null }
        : { time: candles[i].time, value: v - multiplier * std[i]! }
    ),
  };
};

export function formatTimestampKST(unixTime: number): string {
  const date = new Date(unixTime * 1000); // 초 단위 → 밀리초 단위로 변환
  if (isNaN(date.getTime())) return "Invalid time";

  return date.toLocaleString("ko-KR", {
    timeZone: "Asia/Seoul",
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
  });
}

export function calculateVWAP(candles: Candle[]): IndicatorPoint[] {
  const result: IndicatorPoint[] = [];
  let cumulativePV = 0;
  let cumulativeVolume = 0;

  for (let i = 0; i < candles.length; i++) {
    const candle = candles[i];
    const typicalPrice = (candle.high + candle.low + candle.close) / 3;
    cumulativePV += typicalPrice * candle.volume;
    cumulativeVolume += candle.volume;

    const vwap =
      cumulativeVolume === 0 ? null : cumulativePV / cumulativeVolume;
    result.push({ time: candle.time, value: vwap });
  }

  return result;
}
